File: 20. Valid Parentheses/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        parentheses = {")": "(", "]": "[", "}": "{"}
        stack = []

        for char in s:
            if char in parentheses:
                if not stack:
                    return False

                popped = stack.pop()
                if popped != parentheses[char]:
                    return False
            else:
                stack.append(char)

        return len(stack) == 0

File: 20. Valid Parentheses/20. Valid Parentheses.java
import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

class Solution {

  public boolean isValid(String s) {
    Map<Character, Character> parentheses = new HashMap<>();
    parentheses.putIfAbsent(')', '(');
    parentheses.putIfAbsent(']', '[');
    parentheses.putIfAbsent('}', '{');

    Stack<Character> stack = new Stack<>();

    for (char c : s.toCharArray()) {
      if (parentheses.containsKey(c)) {
        if (stack.isEmpty()) {
          return false;
        }

        char popped = stack.pop();
        if (popped != parentheses.get(c)) {
          return false;
        }
      } else {
        stack.add(c);
      }
    }

    return stack.isEmpty();
  }
}
